Enhances the error handling of the heap heuristic

lebr0nli opened the pull request:
This PR enhances the error handling of the heap heuristics.
(The problems are mentioned in https://github.com/pwndbg/pwndbg/issues/1231#issuecomment-1269262842 and https://github.com/pwndbg/pwndbg/issues/1231#issuecomment-1269305915)

And this PR also tries to avoid using `_IO_list_all` to find other symbols if possible.

lebr0nli committed:
Enhance the error handling of the heap command

* Add a new method: `can_be_resolved()` to heap classes to check whether we can resolve the heap after the heap is initialized.

* Add a new function: `get_got_plt_address()` to `pwndbg.glibc`, by doing this, we can determine the location of the symbols without `_IO_list_all` by parsing the output of `info files`.

* Add a new subclass of Exception: `SymbolUnresolvableError` to handle the error when we can't resolve some symbols.

* If we didn't set the GLIBC version manually, we won't get the unnecessary Python error from this now, instead, we tell the user how to set it.

* If we didn't have enough information to resolve some symbols, we show which symbol we lack and tell the user how to set it manually instead of just executing it and showing a Python error.

lebr0nli committed:
Avoid getting the wrong heap config

disconnect3d commented on pwndbg/commands/__init__.py:
```diff
@@ -271,27 +272,35 @@ def _is_statically_linked():
+                )
+        else:
+            print(message.error(f"{function.__name__}: "), end="")
+            if not pwndbg.config.resolve_heap_via_heuristic:
```
if resolve via heuristic is set we wont print anything after func name?

disconnect3d commented on pwndbg/commands/__init__.py:
```diff
@@ -301,19 +310,19 @@ def _OnlyWithResolvedHeapSyms(*a, **kw):
+                    "Please set the GLIBC version you think the target binary was compiled (using `set glibc <version>` command; e.g. 2.32) and re-run this command."
                 )
+            else:
+                e("An unknown error occurred when resolved the heap.")
```
Shall we tell the user we couldn't find proper symbol addresses and they can set them?

lebr0nli replied:
Yes, I'll add the hint for it also.

disconnect3d commented on pwndbg/heap/__init__.py:
```diff
@@ -17,7 +17,7 @@
 )
 
-symbol_list = pwndbg.config.get_params("heap")
+symbol_list = [local for local in locals().values() if isinstance(local, pwndbg.config.Parameter)]
```
Is this just all above or is there more to it? Can we just list it manually? Doing it with `locals()` can be problematic with further code changes

lebr0nli replied:
> Is this just all above or is there more to it?

Yes, we only need above config for now.

> Can we just list it manually?

Yes. (I use `locals()` because we might need to modify this list every time we want to introduce a new symbol to let the user set it, so I want to avoid this, but seems this `locals()` solution is not good haha.)

> Doing it with `locals()` can be problematic with further code changes.

Ok, I'll just list it manually.

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -697,6 +703,9 @@ def libc_has_debug_syms(self):
 
 class DebugSymsHeap(Heap):
+    def can_be_resolved(self):
+        return self.libc_has_debug_syms()
```
Would this work?: `can_be_resolved = Heap.libc_has_debug_syms`

lebr0nli replied:
Yep, it works. I'll change it to your suggestions.

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -908,11 +907,14 @@ def is_glibc_static_symbol(self, addr):
         # Note: This might fail if there is an another symbol also call `_IO_list_all`...
         # TODO/FIXME: Find a better way to check if `addr` is a static symbol of GLIBC.
```
btw what do we mean by 'addr being static symbol in libc'?

lebr0nli replied:
That's mean I want to avoid finding wrong symbol with the same name, I mean:

If we compiled this file with `gcc FILENAME.c -g` or `gcc FILENAME.c -static -g`:

```c
#include <stdlib.h>
void *main_arena = 0xdeadbeaf;
int main(){
    free(malloc(0x20));
    return 0;
}
```

`pwndbg.symbol.address("main_arena")` will return the address of this 0xdeadbeaf `main_arena` instead of the "real" `main_arena` in libc, so the `heap` in debug symbols mode will fail.
But I want to avoid this and make it works in heuristic mode if possible, this why I have this weird check XD.

I'll check if `gdb.lookup_static_symbol` can help later.

lebr0nli committed:
List the symbols manually instead of using `locals()`

lebr0nli committed:
Avoid the extra function call by `can_be_resolved`

lebr0nli committed:
Enhance the error handling when finding TLS (#1237)

lebr0nli committed:
Enhance the error handling for more cases

lebr0nli committed:
Add support to use `gdb.lookup_static_symbol` in `pwndbg.symbol`

lebr0nli committed:
Enhance the strategy when handling the heap-related symbols

* Use `pwndbg.symbol.static_linkage_symbol_address()` to get the address of the symbol first

e.g. Let's say we have a file called `FILENAME.c`:

```
void *main_arena = 0xdeadbeaf;
int main(){
    free(malloc(0x20));
    return 0;
}
```

If we compiled it with `gcc FILENAME.c -g`, the old heap command will fail because it thinks `main_arena` is this 0xdeadbeaf `main_arena`, not the "real" `main_arena` in GLIBC.

With this commit, it should work without this issue.

lebr0nli committed:
Revert "Enhance the error handling when finding TLS (#1237)"

This reverts commit 7d2d1ae6b6cef00d23cc3b8c151dc79862e4c5c3.

lebr0nli committed:
Merge branch 'dev' into update-heap-heuristic

lebr0nli committed:
Enhance the error handling when finding TLS (#1237)

lebr0nli committed:
Catch the error when reading the address of the static linkage symbol

lebr0nli committed:
Bug fix for `thread_cache` under heuristic mode

lebr0nli committed:
Bug fix for `static_linkage_symbol_address()`

* If `gdb.lookup_static_symbol(symbol)` is None, it will cause the
  error.

lebr0nli committed:
Merge branch 'pwndbg:dev' into update-heap-heuristic

lebr0nli committed:
Use new code after refactoring

CptGibbon commented:
Is it just me or does this make debugging a nightmare, since you can no longer use `set exception-verbose on`.

lebr0nli commented:
> Is it just me or does this make debugging a nightmare, since you can no longer use `set exception-verbose on`.

@CptGibbon Ahh, I'm sorry for that, maybe it's because I catch all the exceptions here:

https://github.com/pwndbg/pwndbg/blob/bfbb2b86525ce5c98ef912b47ba0fd75eab4eea7/pwndbg/commands/__init__.py#L286

I'll fix this, I'm really sorry for the inconvenience.

disconnect3d commented:
Huh, great catch folks! At first I didn't understand @CptGibbon issue as it worked for me but with... non-heap commands :P
